File: AÑO 2/programacion_y_diseño_orientado_a_objetos/Practicas/Napakalaki_Ruby/pkg/Napakalaki-0.0.1/lib/player.rb
#encoding: utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative 'dice.rb'
require_relative 'bad_consequence.rb'
require_relative 'treasure.rb'
require_relative 'combat_result.rb'
require_relative 'card_dealer.rb'

class Player
  
  attr_reader :name, :hiddenTreasures, :visibleTreasures, :level, :canISteal
  attr_writer :enemy, :pendingBadConsequence
  @@MAX_LEVEL = 10
  
  def initialize(name)
    @name = name
    @level = 1
    @dead = true
    @canISteal = true
    @hiddenTreasures = Array.new
    @visibleTreasures = Array.new
    @enemy = nil
    @pendingBadConsequence = Array.new
  end
  
  def getHiddenTreasures
    @hiddenTreasures
  end
  
  def getVisibleTreasures
    @visibleTreasures
  end
  
  def maxLevel
    @@MAX_LEVEL
  end
  
  def self.maxLevel
    @@MAX_LEVEL
  end
  
  def bringToLife
    @dead = false
  end
  
  def combatLevel
    bonus = 0
    @visibleTreasures.each do |i|
      bonus = bonus + i.bonus
    end
    (@level + bonus)
  end
  
  def incrementLevels(l)
    @level = [@level + l,@@MAX_LEVEL].min
  end
  
  def decrementLevels(l)
    @level = [@level - l,1].max
  end
  
  def applyPrize(m)
    nLevels = m.getLevelsGained
    incrementLevels(nLevels)
    nTreasures = m.getTreasuresGained
    if(nTreasures > 0)
      dealer = CardDealer.instance
      for i in 1..nTreasures do
        treasure = dealer.nextTreasure
        @hiddenTreasures << treasure
      end
    end
  end
  
  def applyBadConsequence(m)
    badConseq = m.badConsequence
    nLevels = badConseq.levels
    decrementLevels(nLevels)
    pendingBad = badConseq.adjustToFitTreasureLists(@visibleTreasures,@hiddenTreasures)
    puts pendingBad #Quitar luego
    @pendingBadConsequence = pendingBad
  end
  
  def canMakeTreasureVisible(t)
    canMakeTVisible = true
    types = Array.new
    hands = 0
    
    for treasure in @visibleTreasures do
      types << treasure.type
      if(treasure.type == TreasureKind::ONEHAND)
        hands = hands + 1
      elsif(treasure.type == TreasureKind::BOTHHANDS)
        hands = hands + 2
      end
    end
    
    kind = t.type
    if(kind == TreasureKind::ONEHAND)
      hands = hands + 1
    elsif(kind == TreasureKind::BOTHHANDS)
      hands = hands + 2
    end
    
    if(types.include?(kind) && kind != TreasureKind::ONEHAND && kind != TreasureKind::BOTHHANDS)
      canMakeTVisible = false
    else
      canMakeTVisible = true
    end
    
    canMakeTVisible && (hands <= 2)
  end
  
  def howManyVisibleTreasures(tKind)
    numberOftKind = 0
    @visibleTreasures.each do |i|
      if(i.type == tKind)
        numberOftKind += 1
      end
    end
    numberOftKind
  end
  
  def dieIfNoTreasures
    if(@visibleTreasures.empty? && @hiddenTreasures.empty?) then
      @dead = true
    end
  end
  
  def isDead #Pensar en cambiar por un attr_reader
    @dead
  end
  
  def combat(m)
    myLevel = combatLevel
    monsterLevel = m.combatLevel
    cr = nil
    if(!@canISteal) then
      dice = Dice.instance
      number = dice.nextNumber
      if(number < 3) then
        enemyLevel = enemy.combatLevel
        monsterLevel = monsterLevel + enemyLevel
      end
    end
    if(myLevel > monsterLevel) then
      applyPrize(m)
      cr = (@level >= maxLevel)? CombatResult::WINGAME : CombatResult::WIN
    else
      applyBadConsequence(m)
      cr = CombatResult::LOSE
    end
    cr
  end
  
  def makeTreasureVisible(t)
    canI = canMakeTreasureVisible(t)
    if(canI) then
      @visibleTreasures << t
      @hiddenTreasures.delete(t)
    end
  end
  
  def discardVisibleTreasure(t)
    @visibleTreasures.delete(t)
    if((@pendingBadConsequence != nil) && !@pendingBadConsequence.empty?) then
      @pendingBadConsequence.substractVisibleTreasure(t)
    end
    dieIfNoTreasures
  end
  
  def discardHiddenTreasure(t)
    @hiddenTreasures.delete(t)
    if((@pendingBadConsequence != nil) && !@pendingBadConsequence.empty?) then
      @pendingBadConsequence.substractHiddenTreasure(t)
    end
    dieIfNoTreasures
  end
  
  def validState
    return ((@hiddenTreasures.size < 5) && (@pendingBadConsequence.empty?));
  end
  
  def initTreasures
    dealer = CardDealer.instance
    dice = Dice.instance
    bringToLife
    #Primer tesoro
    @hiddenTreasures << dealer.nextTreasure
    number = dice.nextNumber
    #Segundo tesoro
    if(number > 1) then
      @hiddenTreasures << dealer.nextTreasure
    end
    #Tercer tesoro
    if(number == 6) then
      @hiddenTreasures << dealer.nextTreasure
    end
  end

  def stealTreasure
    treasure = nil
    canI = canISteal
    if(canI) then
      canYou = @enemy.canYouGiveMeATreasure
      if(canYou) then
        treasure = @enemy.giveMeATreasure
        @hiddenTreasures << treasure
        haveStolen
      end
    end
    treasure
  end
  
  def giveMeATreasure
    n = @hiddenTreasures.size
    @hiddenTreasures[rand(n)]
  end
  
  def canYouGiveMeATreasure
    return !@hiddenTreasures.empty?
  end
  
  def haveStolen
    @canISteal = false
  end
  
  def discardAllTreasures
    cVisibleTreasures = Array.new(@visibleTreasures)
    cHiddenTreasures = Array.new(@hiddenTreasures)
    for t in cVisibleTreasures do
      discardVisibleTreasure(t)
    end
    for t in cHiddenTreasures do
      discardHiddenTreasure(t)
    end
  end
  
  def to_s
    "#{@name}\nlevel=#{@level}, combatLevel=#{combatLevel}, dead=#{@dead}, enemy=#{@enemy.name}";
  end
  
  private :bringToLife, :combatLevel, :incrementLevels, :decrementLevels, :applyPrize, :applyBadConsequence, :canMakeTreasureVisible, :howManyVisibleTreasures, :dieIfNoTreasures, :giveMeATreasure, :canYouGiveMeATreasure, :haveStolen
  
end

File: AÑO 2/programacion_y_diseño_orientado_a_objetos/Practicas/Napakalaki_Java/src/NapakalakiGame/Player.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import java.util.ArrayList;
import java.util.Random;
import GUI.Dice;

/**
 *
 * @author guillermo
 */
public class Player {
    final static int MAXLEVEL = 10;
    private String name;
    private int level;
    private boolean dead = true;
    private boolean canISteal = true;
    protected Player enemy;
    private ArrayList<Treasure> hiddenTreasures = new ArrayList();
    private ArrayList<Treasure> visibleTreasures = new ArrayList();
    private BadConsequence pendingBadConsequence;
    
    public Player(String name){
        this.name = name;
        this.level = 1;
    }
    
    public Player(Player p){
        this.name = p.name;
        this.level = p.level;
        this.dead = p.dead;
        this.canISteal = p.canISteal;
        this.enemy = p.enemy;
        this.hiddenTreasures = p.hiddenTreasures;
        this.visibleTreasures = p.visibleTreasures;
        this.pendingBadConsequence = p.pendingBadConsequence;
    }
    
    public String getName(){
        return name;
    }
    
    protected int getOponentLevel(Monster m){
        return m.getCombatLevel();
    }
    
    protected boolean shouldConvert(){
        boolean shouldconvert = false;
        Dice dice = Dice.getInstance();
        if (dice.nextNumber() == 6) {
            shouldconvert = true;
        }
        return shouldconvert;
    }
    
    public boolean isACultistPlayer(){
        return false;
    }
    
    private void bringToLife(){
        dead = false;
    }
    
    protected int getCombatLevel(){
        int flevel = level;
        for(Treasure t: visibleTreasures)
            flevel += t.getBonus();
        return flevel;
    }
    
    private void incrementLevel(int i){
        level = Math.min(level + i, MAXLEVEL);
    }
    
    private void decrementLevel(int i){
        level = Math.max(level-i,1);
    }
    
    private void setPendingBadConsequence(BadConsequence bc){
        pendingBadConsequence = bc;
        //Esto no pertenece al juego, está así por jugabilidad :)
        System.out.println(pendingBadConsequence.toString());
    }
    
    private void applyPrize(Monster m){
        int nLevels = m.getLevelsGained();
        incrementLevel(nLevels);
        int nTreasures = m.getTreasuresGained();
        if(nTreasures>0){
            CardDealer dealer = CardDealer.getInstance();
            for(int i=0; i<nTreasures; i++){
                Treasure treasure = dealer.nextTreasure();
                hiddenTreasures.add(treasure);
            }
        }
    }
    
    private void applyBadConsequence(Monster m){
        BadConsequence badConsequence = m.getBadConsequence();
        
        int nLevels = badConsequence.getLevels();
        decrementLevel(nLevels);
        BadConsequence pendingBad = badConsequence.adjustToFitTreasureLists(visibleTreasures, hiddenTreasures);
        setPendingBadConsequence(pendingBad);
    }
    
    private boolean canMakeTreasureVisible(Treasure t){
        boolean canMakeTreasureVisible = true;
        ArrayList<TreasureKind> types = new ArrayList();
        int hands = 0;
        
        for(Treasure treasure:visibleTreasures){
            types.add(treasure.getType());
            if(treasure.getType() == TreasureKind.ONEHAND)
                hands++;
            else if(treasure.getType() == TreasureKind.BOTHHANDS)
                hands += 2;
        }
        TreasureKind kind = t.getType();
        if(kind == TreasureKind.ONEHAND){
            hands++;
        }
        else if(kind == TreasureKind.BOTHHANDS){
            hands+=2;
        }

        if(types.contains(kind) && kind != TreasureKind.ONEHAND && kind != TreasureKind.BOTHHANDS){
            canMakeTreasureVisible = false;
        }
        else{
            canMakeTreasureVisible = true;
        }
        
        return canMakeTreasureVisible && (hands <=2);
    }
    
    private int howManyVisibleTreasures(TreasureKind tKind){
        int res = 0;
        for(Treasure r: visibleTreasures)
            if(r.getType() == tKind)
                res++;
        return res;
    }
    
    private void dieIfNoTreasures(){
        if(hiddenTreasures.isEmpty() && visibleTreasures.isEmpty())
            dead = true;
    }
    
    public boolean isDead(){return dead;}
    
    public ArrayList<Treasure> getHiddenTreasures(){return hiddenTreasures;} //Hay que confirmar esto
    
    public ArrayList<Treasure> getVisibleTreasures(){return visibleTreasures;} //Hay que confirmar esto
    
    public CombatResult combat(Monster m){
        int myLevel = getCombatLevel();
        int oponentLevel = getOponentLevel(m);
        CombatResult cr;
        if(!canISteal){
            Dice dice = Dice.getInstance();
            int number = dice.nextNumber();
            if(number<3){
                int enemyLevel = enemy.getCombatLevel();
                oponentLevel += enemyLevel;
            }
        }
        if(myLevel>oponentLevel){
            applyPrize(m);
            cr=(this.level >= MAXLEVEL)?CombatResult.WINGAME:CombatResult.WIN;
        }
        else{
            applyBadConsequence(m);
            cr=CombatResult.LOSE;
            if(shouldConvert())
                cr = CombatResult.LOSEANDCONVERT;
        }
        return cr;
    }
    
    public void makeTreasureVisible(Treasure t){
        boolean canI = canMakeTreasureVisible(t);
        if(canI){
            visibleTreasures.add(t);
            hiddenTreasures.remove(t);
        }
    }
    
    public void discardVisibleTreasure(Treasure t){
        visibleTreasures.remove(t);
        if((pendingBadConsequence != null) && !pendingBadConsequence.isEmpty())
            pendingBadConsequence.substractVisibleTreasure(t);
        dieIfNoTreasures();
    }
    
    public void discardHiddenTreasure(Treasure t){
        hiddenTreasures.remove(t);
        if((pendingBadConsequence != null) && !pendingBadConsequence.isEmpty())
            pendingBadConsequence.substractHiddenTreasure(t);
        dieIfNoTreasures();
    }
    
    public boolean validState(){
        return (hiddenTreasures.size() < 5 && (pendingBadConsequence == null || pendingBadConsequence.isEmpty()));
    }
    
    public void initTreasures(){
        CardDealer dealer = CardDealer.getInstance();
        Dice dice = Dice.getInstance();
        bringToLife();
        //Primer tesoro
        hiddenTreasures.add(dealer.nextTreasure());
        int number = dice.nextNumber();
        //Segundo tesoro
        if(number>1)
            hiddenTreasures.add(dealer.nextTreasure());
        //Tercer tesoro
        if(number==6)
            hiddenTreasures.add(dealer.nextTreasure());
    }
    
    public int getLevels(){
        return level;
    }
    
    public Treasure stealTreasure(){
        Treasure treasure = null;
        boolean canI = canISteal();
        if(canI){
            boolean canYou = enemy.canYouGiveMeATreasure();
            if(canYou){
                treasure = enemy.giveMeATreasure();
                hiddenTreasures.add(treasure);
                haveStolen();
            }
        }
        return treasure;
    }
    
    public void setEnemy(Player enemy){
        this.enemy = enemy;
    }
    
    private Treasure giveMeATreasure(){
        int n = hiddenTreasures.size();
        Random rand = new Random();
        return hiddenTreasures.get(rand.nextInt(n));
    }
    
    public boolean canISteal(){
        return canISteal;
    }
    
    private boolean canYouGiveMeATreasure(){
        return !hiddenTreasures.isEmpty();
    }
    
    private void haveStolen(){
        canISteal = false;
    }
    
    public void discardAllTreasures(){
        ArrayList<Treasure> cVisibleTreasures = new ArrayList(visibleTreasures);
        ArrayList<Treasure> cHiddenTreasures = new ArrayList(hiddenTreasures);
        for(Treasure t:cVisibleTreasures)
            discardVisibleTreasure(t);
        for(Treasure t:cHiddenTreasures)
            discardHiddenTreasure(t);
    }
    
    public String getEnemyName(){
        return enemy.getName();
    }

    protected Player getEnemy(){
        return enemy;
    }

    public BadConsequence getPendingBadConsequence() {
        return pendingBadConsequence;
    }
    
    @Override
    public String toString() {
        return name + ":\n nivel=" + level + ", nivel de combate= " + getCombatLevel() + ", muerto=" + dead + ", enemigo=" + enemy.name;
    }


}
